core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -2380,6 +2381,8 @@ public static boolean classifyFilters(
         joinRel.getInputs().get(1).getRowType().getFieldList();
     final int nFieldsRight = rightFields.size();
     assert nTotalFields == (joinRel instanceof SemiJoin
```

Comment:
If I am not mistaken nSysFields is not used anymore and is always zero. If that's the case can you remove it please. Moreover the assertion looks a bit weird (plus a bit trivial). If the method cannot treat all kinds of joins then that should be a precondition with an IllegalArgumentException and not an assertion. I would be inclined to remove entirely this assertion.

Reply:
I agree, this assert starts to become too complex (a sort of "switch" to detect all the special cases where the joinRel generates only the left side fields) to verify something that is not the responsibility of this static method. Probably it would be better to just remove it.

Reply:
The goal of the assert is to educate future readers and maintainers of this code. In this case, try refactoring the assert to improve clarity.

The maintainer of the code does need to know about all that complexity (in particular how complex the type hierarchy of Join, Correlate and SemiJoin has gotten). Removing the assert is just sweeping the issue under the rug. We should be embarrassed about the complexity and strive to fix it.

Reply:
(I just discovered that `Correlate`, which is neither a Join nor a SemiJoin, uses `SemiJoinType`, but `SemiJoin` does not use `SemiJoinType`. Yuck.)

Reply:
I totally agree that the hierarchy of Correlate, Join, etc., needs to be redesigned but I don't think this can/should be done as part of this issue.

Regarding the assertion, I think it is a bit misplaced; it verifies how many fields the join has in relation with its inputs (depending also on the type). Such an assertion would appear reasonable if it was introduced when the respective operators are created but not in a method which is a mere consumer of the operator.

Reply:
Hey @julianhyde, what do you think of my previous comment? Do you still think we should keep the assertion here?

Reply:
I think the assert is really useful. It tells the maintainer of the code what they are dealing with.

It's ugly because we have a mess to clean up. Leave it until we've cleaned up the mess.

If we wanted an invariant, then we'd use `Objects` or `Preconditions` methods that, unlike `assert`, do not disappear, and we'd add them to the constructor.

Reply:
Thanks all for your comments. The purpose of this PR is just to fix the IllegalStateException in FilterCorrelateRule.java, so I'd say let's keep the assert, and I'll squash the commits so this can be merged.

Reply:
Fine by me although I have a small remark. The way you modified the code right now the if statements are going to be executed even if assertions are disabled. I don't know if JIT is intelligent enough to remove entirely the redundant if but I don't think we should rely on that. Furthemore, to make it more readable you could possibly extra the check into a method, but feel free to modify as you see fit I don't feel strongly about it.

Reply:
Ok, I'll refactor the code to use an auxiliary method to check the condition